picard/acousticbrainz/__init__.py

```diff
@@ -0,0 +1,130 @@
+                          ACOUSTICBRAINZ_PORT,
+                          "/%s/low-level" % recording_id,
+                          partial(run_extractor, input_path, output_path, extractor_callback),
+                          priority=True,
+                          important=False,
+                          parse_response_type=None)
+
+
+def extractor(input_path, output_path):
+    # If an a duplicate entry doesn't exist, process the file
+    process = QtCore.QProcess()
+    process.setProperty('picard_finished', False)
+    process.setProcessChannelMode(QtCore.QProcess.MergedChannels)  # Merge stdout and stderr
+    process.start(acousticbrainz_extractor, [input_path, output_path])
+    process.waitForFinished(msecs=-1)  # Do not timeout
+    results = (output_path, process.exitCode(), process.readAllStandardOutput().data().decode('utf-8'))
+
+    # Add feature extractor sha to the output file
+    try:
+        with open(output_path, "r", encoding="utf-8") as f:
+            features = json.load(f)
+        features["metadata"]["version"]["essentia_build_sha"] = acousticbrainz_extractor_sha
+        with open(output_path, "w", encoding="utf-8") as f:
+            json.dump(features, f, indent=4)
+    except FileNotFoundError:
+        pass
+
+    return results
+
+
+def run_extractor(input_path, output_path, extractor_callback, response, reply, error):
+    duplicate = False
+    # Check if AcousticBrainz server answered with the json file for the recording id
+    if not error:
+        try:
+            load_json(response)
+            duplicate = True
+        except json.JSONDecodeError:
+            pass
+
+    if duplicate:
+        # If an entry for the same recording ID already exists
+        # using the same extractor version exists, skip extraction
+        results = (None, 0, "Duplicate")
+        extractor_callback(results, None)
+    else:
+        # Run extractor on a different thread and call the callback when done
+        run_task(partial(extractor, input_path, output_path), extractor_callback)
+
+
+def submission_callback(tagger, file, data, http, error):
+    if not error:
+        file.acousticbrainz_is_duplicate = True
+        tagger.log.debug("AcousticBrainz features were successfully submitted: %s" % file.filename)
+    else:
+        tagger.log.debug("AcousticBrainz features were not submitted: %s" % file.filename)
+
+
+def submit_features(tagger, file, features_file):
+    with open(features_file, "r", encoding="utf-8") as f:
+        features = json.load(f)
+
+    # Check if extracted recording id matches the current file
+    if features['metadata']['tags']['musicbrainz_trackid'] != file.metadata['musicbrainz_recordingid']:
+        tagger.log.debug("AcousticBrainz features recordingId do not match the file metadata: %s" % file.filename)
+        submission_callback(tagger, file, None, None, True)
+        return
+
+    featstr = json.dumps(features)
+
+    tagger.webservice.post(ACOUSTICBRAINZ_HOST,

```

Comment:
Reduce indentation of parameters by wrapping just after first parenthesis and just before last one:

```python
tagger.webservice.post(
    ACOUSTICBRAINZ_HOST,
    ACOUSTICBRAINZ_PORT,
    "/%s/low-level" % file.metadata["musicbrainz_recordingid"],
...
)
```

Reply:
Done.